### PR TITLE
Set up a rolling CI test run one week from "now"

### PR DESCRIPTION
**Pull request by elliotcm**
## Context

We've had problems with timebomb tests which break at different part of the cycles.  This allows us to anticipate these issues by giving us one week's notice, and lays the groundwork for more comprehensive or sophisticated setups later on.

## Link to Trello card

https://trello.com/c/dgbZ5UaJ/753-allow-running-the-test-suite-at-an-arbitrary-date

### CODE REVIEW

**Comment by elliotcm** [comment]
[Allow running the test suite at an arbitrary date](https://trello.com/c/dgbZ5UaJ/753-allow-running-the-test-suite-at-an-arbitrary-date)

**Comment by carlosmartinez** [comment]
this is a cool idea! could you walk the dev team through it tomorrow?

**Comment by elliotcm** [comment]
> this is a cool idea! could you walk the dev team through it tomorrow?

@carlosmartinez yep

File: adr/0020-strict-time-control-in-test-suite.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+- We will reveal any code which is unwittingly dependent on e.g. the precision of timestamps and/or at risk of race conditions
+- We can externally control the presumed time of the test suite, which will allow us to run the test suite against multiple important dates and deadlines, as well as sweep ahead for time-sensitive tests by running the suite at some rolling period in the future
+- We remove a coupling with `Timecop`
+
+### Cons
+
+- We will need to update a lot of tests to use the new API
+- We replace the `Timecop` coupling with a new (less coupled) coupling on our own `TestSuiteTimeMachine` wrapper, which may itself be extracted to a gem in the future
+- We will have to think more carefully about the effects of time on our tests as we write them
+- We will need to write code which makes no assumptions about the passage of time in relation to e.g. the creation of records
```
Comment (tomas-stefano) [inline adr/0020-strict-time-control-in-test-suite.md:51]: ❤️ 🚀

**Review by tomas-stefano — APPROVED** [review]

File: app/models/support_interface/vendor_api_monitor.rb [hunk 2]
```diff
@@ -26,21 +26,21 @@ def no_sync_in_24h
       connected
         .select('last_syncs.last_sync as last_sync, vendor_id')
         .joins("LEFT JOIN (#{VendorAPIRequest.successful.syncs.select('provider_id, MAX(vendor_api_requests.created_at) as last_sync').group('provider_id').to_sql}) last_syncs on last_syncs.provider_id = providers.id")
-        .where("last_sync < now() - interval '24 hours' OR last_sync IS NULL").order('last_sync DESC')
+        .where("last_sync < ('#{Time.zone.now.utc.iso8601(6)}'::timestamp - interval '24 hours') OR last_sync IS NULL").order('last_sync DESC')
```
Comment (carlosmartinez) [inline app/models/support_interface/vendor_api_monitor.rb:29]: should we extract this to `postgres_current_time` or similar?

**Review by carlosmartinez — APPROVED** [review]

File: lib/test_suite_time_machine.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    if @baseline_set
+      raise TimeTravelError, 'TestSuiteTimeMachine.pretend_it_is cannot be called more than once per test run'
+    end
+
+    datetime = 'real_world' if datetime.blank?
+
+    Timecop.baseline =  case datetime
+                        when 'real_world'
+                          Time.zone.now
+                        when /\A(\d+).days.from_now\z/
```
[the remark this answers is not in the document]
Comment (elliotcm) [reply lib/test_suite_time_machine.rb:12]: This is the one being used by CI.

File: app/models/support_interface/vendor_api_monitor.rb (hunk 2 above)
Comment (elliotcm) [reply app/models/support_interface/vendor_api_monitor.rb]: Done (pending test fixes)

File: lib/test_suite_time_machine.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (carlosmartinez) [reply lib/test_suite_time_machine.rb:12]: gotcha, thanks

**Pull request merged**